presto-main/src/main/java/io/prestosql/operator/JoinProbe.java

```diff
@@ -120,8 +120,22 @@ private void fillJoinPositionCache(LookupSource lookupSource)
         // Extracted to local variables for performance reasons
         int firstPosition = this.position & (~JOIN_POSITIONS_CACHE_MASK);
         int limit = min(JOIN_POSITIONS_CACHE_SIZE, positionCount - firstPosition);
-        for (int i = 0; i < limit; ++i) {
-            joinPositionsCache[i] = getJoinPosition(firstPosition + i, lookupSource);
+        if (mayContainNullRows()) {
+            for (int i = 0; i < limit; ++i) {
+                joinPositionsCache[i] = getJoinPosition(firstPosition + i, lookupSource);
+            }
+        }
+        else {
+            if (probeHashBlock.isPresent()) {
+                long[] rawHashes = new long[limit];
+                for (int i = 0; i < limit; ++i) {
+                    rawHashes[i] = BIGINT.getLong(probeHashBlock.get(), firstPosition + i);
```

Comment:
you can pass `Block rawHashes` instead. You can use `Block#getRegion` (it returns view on block)

Reply:
Unfortunately it increases time by ~7%. Probably due to bound checks everywhere